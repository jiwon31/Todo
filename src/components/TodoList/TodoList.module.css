.container {
  height: 100%;
  min-height: 0;
  display: flex;
  flex-direction: column;
  background-color: var(--color-list-bg);
}

.list {
  flex: 1 1 0;
  overflow-y: auto;
}
