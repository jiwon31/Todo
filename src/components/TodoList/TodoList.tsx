import { Filter } from "App";
import AddForm from "components/AddForm/AddForm";
import TodoItem from "components/TodoItem/TodoItem";
import { useState, useEffect } from "react";
import styles from "./TodoList.module.css";
import TodoPresenter, { Todo } from "./todo_presenter";

type TodoListProps = {
  present: TodoPresenter;
  filter: Filter;
};

export default function TodoList({ present, filter }: TodoListProps) {
  /* ⭐ 중요 포인트!
  컴포넌트가 리렌더링될 때마다 useState도 다시 호출이 되어서 초기값이 다시 전달되는데, useState 내부적으로 저장된 값이 있다면 초기값을 무시하고 내부적으로 사용하고 있는 값을 사용한다.
  함수를 호출해서 데이터를 읽어오는 경우(로컬스토리지나 파일, 네트워크 통신 등), 컴포넌트가 리렌더링될 때마다 계속 데이터를 읽어온다.
  그리고 내부적으로 다시 읽어온 값을 쓰지 않고 원래 가지고 있는 값을 사용한다. UI상으로 업데이트 되지 않아도 계속 불필요하게 초기값을 읽어오게 된다.
  초기값으로 함수를 호출하는 경우라면(특히, 함수 안에서 무거운 일을 하는 경우라면) 함수를 호출한 값 자체를 전달하는 것이 아니라 콜백함수로 한 단계 감싸줘야 한다.
  => 초기값이 필요한 경우에만 전달한 함수를 호출해서 초기값을 얻는다. */
  const [todos, setTodos] = useState<Todo[]>(() => present.getTodos());

  const handleAdd = (added: Todo) => present.add(added, setTodos);
  const handleDelete = (deleted: Todo) => present.delete(deleted, setTodos);
  const handleUpdate = (updated: Todo) => present.update(updated, setTodos);

  /* ✅ useEffect
  1. 특정한 값이 변경될 때마다 무언가 하고 싶다면
    리액트에서 state가 변경이 되어서 UI를 업데이트하는 흐름이 아니라 side effect으로(부가적으로), state가 변경이 되어서 UI가 업데이트될 때마다 추가적으로 해줘야되는게 있다면
  2. 컴포넌트가 마운트될 때 처음으로 해줘야될 것이 있다면
  3. 컴포넌트가 언마운트될 때 정리해줘야될 것이 있다면  */
  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const filterd = present.getFilteredTodos(filter);
  return (
    <section className={styles.container}>
      <ul className={styles.list}>
        {filterd.map((item) => (
          <TodoItem
            key={item.id}
            todo={item}
            onDelete={handleDelete}
            onUpdate={handleUpdate}
          />
        ))}
      </ul>
      <AddForm onAdd={handleAdd} />
    </section>
  );
}
