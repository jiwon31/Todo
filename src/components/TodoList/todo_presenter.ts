import { Filter } from "App";

export interface TodoPresenter {
  getTodos(): Todo[];
  add(added: Todo, update: UpdateTodo): void;
  delete(deleted: Todo, update: UpdateTodo): void;
  update(updated: Todo, update: UpdateTodo): void;
  getFilteredTodos(filter: Filter): Todo[];
}

export type Todo = {
  id: string;
  text: string;
  status: "active" | "completed";
};
type UpdateTodo = (todos: Todo[]) => void;

export class TodoPresenterImpl implements TodoPresenter {
  constructor(private todos: Todo[]) {}

  getTodos(): Todo[] {
    return this.todos;
  }

  add(added: Todo, update: UpdateTodo) {
    this.todos = [...this.todos, added];
    update(this.todos);
  }

  delete(deleted: Todo, update: UpdateTodo) {
    this.todos = this.todos.filter((todo) => todo.id !== deleted.id);
    update(this.todos);
  }

  update(updated: Todo, update: UpdateTodo) {
    this.todos = this.todos.map((todo) =>
      todo.id === updated.id ? updated : todo
    );
    update(this.todos);
  }

  getFilteredTodos(filter: Filter): Todo[] {
    if (filter === "all") {
      return this.todos;
    }
    return this.todos.filter((todo) => todo.status === filter);
  }
}
