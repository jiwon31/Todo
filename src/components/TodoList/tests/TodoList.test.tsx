import TodoList from "../TodoList";
import { TodoPresenter, TodoPresenterImpl } from "../todo_presenter";
import renderer from "react-test-renderer";

describe("TodoList", () => {
  let present: TodoPresenter;
  beforeEach(() => {
    present = new TodoPresenterImpl([
      { id: "280a", text: "studying", status: "active" },
      { id: "280b", text: "running", status: "completed" },
    ]);
  });

  it("renders", () => {
    const component = renderer.create(
      <TodoList present={present} filter={"all"} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
