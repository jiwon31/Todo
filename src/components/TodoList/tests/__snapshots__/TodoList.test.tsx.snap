// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders 1`] = `
<section
  className="container"
>
  <ul
    className="list"
  >
    <li
      className="todo"
    >
      <input
        checked={false}
        className="checkbox"
        data-testid="checkbox"
        id="280a"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="text false"
        data-testid="todo-text"
        htmlFor="280a"
      >
        studying
      </label>
      <button
        className="button"
        onClick={[Function]}
        title="delete"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 1 0z"
          />
        </svg>
      </button>
    </li>
    <li
      className="todo"
    >
      <input
        checked={true}
        className="checkbox"
        data-testid="checkbox"
        id="280b"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="text completed"
        data-testid="todo-text"
        htmlFor="280b"
      >
        running
      </label>
      <button
        className="button"
        onClick={[Function]}
        title="delete"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 1 0z"
          />
        </svg>
      </button>
    </li>
  </ul>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="Add Todo"
      type="text"
      value=""
    />
    <button
      className="button"
    >
      Add
    </button>
  </form>
</section>
`;
